### PR TITLE
Feature/add bml cli commands to dynamic channel selection bml apis

### PR DESCRIPTION
**Pull request by adam1985d**
This PR adds CLI BML commands to the beerocks_cli for the following DCS BML APIS:
bml_set_dcs_continuous_scan_enable
bml_get_dcs_continuous_scan_enable
bml_set_dcs_continuous_scan_params
bml_get_dcs_continuous_scan_params
bml_start_dcs_single_scan
bml_get_dcs_scan_results

The commands naming are aligned to the called APIs without the 'bml_' initial:
set_dcs_continuous_scan_enable
get_dcs_continuous_scan_enable
set_dcs_continuous_scan_params
get_dcs_continuous_scan_params
start_dcs_single_scan
get_dcs_scan_results

The command's description is available in the beerocks_cli help menu

### CODE REVIEW

**Comment by adam1985d** [comment]
[add cli to all DCS BMLs](https://trello.com/c/GrioG4Rj/80-add-cli-to-all-dcs-bmls)

File: controller/src/beerocks/cli/beerocks_cli_bml.cpp
Comment (morantr) [inline controller/src/beerocks/cli/beerocks_cli_bml.cpp:1832]: why the `enable` is `int8` and not `bool`?

File: controller/src/beerocks/cli/beerocks_cli_bml.cpp
Comment (morantr) [inline controller/src/beerocks/cli/beerocks_cli_bml.cpp]: I prefer using types that imply the number of bits that are used: uint32_t/int32_t.

apply this on all new stubs.

File: controller/src/beerocks/cli/beerocks_cli_bml.h
Comment (morantr) [inline controller/src/beerocks/cli/beerocks_cli_bml.h]: I don't know think it is a good idea to define too many optional parameters, especially when they share common types.
I think a better idea will be to define a struct with default parameters and give it an argument.
(nitpick)

Comment (morantr) [inline controller/src/beerocks/cli/beerocks_cli_bml.h]: There is no reason to add here `channel_pool_size` since it is a user interface. The function should automatically figure how many channels there are in the user input.

File: controller/src/beerocks/cli/beerocks_cli_bml.h
Comment (morantr) [inline controller/src/beerocks/cli/beerocks_cli_bml.h]: Why are you passing `channel_pool` as a string and not a vector or array? The caller function should parse it and convert it to vector/array.

File: controller/src/beerocks/cli/beerocks_cli_bml.cpp
Comment (morantr) [inline controller/src/beerocks/cli/beerocks_cli_bml.cpp]: why the` is_single_scan` is int8 and not bool?

File: controller/src/beerocks/cli/beerocks_cli_bml.cpp
Comment (morantr) [inline controller/src/beerocks/cli/beerocks_cli_bml.cpp]: You can define it like this only if you set all the optional arguments altogether. This is not the case here, so what you need to do is to explicitly write to what argument do you mean:
`<mac> <dwell_time_msec> [ interval_time=<interval_time>] [channel_pool=<channel_pool>]`

on the caller function you will need to parse it, and here convert all the argument types to `STRING_ARG`.

File: controller/src/beerocks/cli/beerocks_cli_bml.cpp
Comment (tomereli) [inline controller/src/beerocks/cli/beerocks_cli_bml.cpp]: Probably better to also print the radio mac and the enable.
Can we use `bool enable` instead of `int8_t`?

Same comments for the rest of the APIs

File: controller/src/beerocks/cli/beerocks_cli_bml.h
Comment (tomereli) [inline controller/src/beerocks/cli/beerocks_cli_bml.h:199]: Please add doxygen to all new APIs either here or in the implementation file.
A rule of thumb - every new function comes with doxygen comments.

File: controller/src/beerocks/cli/beerocks_cli_bml.cpp
Comment (tomereli) [inline controller/src/beerocks/cli/beerocks_cli_bml.cpp]: Why don't we print such an error in other failure cases?

File: controller/src/beerocks/cli/beerocks_cli_bml.cpp
Comment (tomereli) [inline controller/src/beerocks/cli/beerocks_cli_bml.cpp]: No need for `struct` keyword - its CPP.

File: controller/src/beerocks/cli/beerocks_cli_bml.cpp
Comment (tomereli) [inline controller/src/beerocks/cli/beerocks_cli_bml.cpp]: Did you run KW on this? please refrain from using `if (ptr == nullptr)` and instead use `if (!ptr)` to avoid KW issues.

File: controller/src/beerocks/cli/beerocks_cli_bml.cpp
Comment (tomereli) [inline controller/src/beerocks/cli/beerocks_cli_bml.cpp]: I'm not sure I follow - what is temporary with the debug print? Aren't the results already in the new structure?

File: controller/src/beerocks/cli/beerocks_cli_bml.cpp
Comment (tomereli) [inline controller/src/beerocks/cli/beerocks_cli_bml.cpp]: There is absolutely no good reason to use `calloc` or `alloc` in c++, and especially in this function where you simply print the returned results and free them right after.
Simply declare the results on the stack:

`BML_DCS_NEIGHBOR_AP results[max_results_size];`

Also - why is the `max_results_size` argument even needed? This can also be statically declared in the function.

**Review by tomereli — CHANGES_REQUESTED** [review]

File: controller/src/beerocks/cli/beerocks_cli_bml.cpp
Comment (morantr) [inline controller/src/beerocks/cli/beerocks_cli_bml.cpp]: string_utils::bool_str(enable);

File: controller/src/beerocks/cli/beerocks_cli_bml.cpp
Comment (morantr) [inline controller/src/beerocks/cli/beerocks_cli_bml.cpp]: Why didn't you declare it on the stack???
Anyway, this is C allocation. The CPP way is with `new`/`delete` and the better option will be to use smart pointers.

**Review by morantr — CHANGES_REQUESTED** [review]

File: controller/src/beerocks/cli/beerocks_cli_bml.cpp
Comment (adam1985d) [reply controller/src/beerocks/cli/beerocks_cli_bml.cpp]: this is a special case where 2 params must be provided together or omitted together

File: controller/src/beerocks/cli/beerocks_cli_bml.cpp
Comment (adam1985d) [reply controller/src/beerocks/cli/beerocks_cli_bml.cpp]: 1. bool instead of uint8_t:
- the BML API uses int - and the CML CLI provides an int - so I don't see a value in int->bool->int conversion
- in the 'get_dcs_scan_results()' I will move the int->bool translation to the caller function and update the signature
2. print the input values:
done

File: controller/src/beerocks/cli/beerocks_cli_bml.h
Comment (adam1985d) [reply controller/src/beerocks/cli/beerocks_cli_bml.h:199]: I added inside the CPP file (not to clatter the class header)

File: controller/src/beerocks/cli/beerocks_cli_bml.cpp
Comment (adam1985d) [reply controller/src/beerocks/cli/beerocks_cli_bml.cpp]: removed the struct keyword

File: controller/src/beerocks/cli/beerocks_cli_bml.cpp
Comment (adam1985d) [reply controller/src/beerocks/cli/beerocks_cli_bml.cpp]: no - since the PR that changes the structure is not available yet
(PRs 4/5): 
https://github.com/prplfoundation/prplMesh/pull/694

File: controller/src/beerocks/cli/beerocks_cli_bml.cpp
Comment (adam1985d) [reply controller/src/beerocks/cli/beerocks_cli_bml.cpp]: the max_results_size is to simulate the max-results-size limit also used in <wave-api> in RDKB (currently set as 200)
when DCS supports ordering of the results by RSSI or another ranking, it is possible to the get only the X best results even if more are available

I replaced the <calloc> with a stack declaration.

File: controller/src/beerocks/cli/beerocks_cli_bml.cpp
Comment (adam1985d) [reply controller/src/beerocks/cli/beerocks_cli_bml.cpp:1832]: the BML API uses int - and the CML CLI provides an int - so I don't see a value in int->bool->int conversion

File: controller/src/beerocks/cli/beerocks_cli_bml.cpp
Comment (adam1985d) [reply controller/src/beerocks/cli/beerocks_cli_bml.cpp]: done

File: controller/src/beerocks/cli/beerocks_cli_bml.h
Comment (adam1985d) [reply controller/src/beerocks/cli/beerocks_cli_bml.h]: might be, but will not be changed in this PR

Comment (adam1985d) [reply controller/src/beerocks/cli/beerocks_cli_bml.h]: discussed this with @tomereli  and he agrees with you. will be removed

File: controller/src/beerocks/cli/beerocks_cli_bml.h
Comment (adam1985d) [reply controller/src/beerocks/cli/beerocks_cli_bml.h]: none of the caller functions in the BML CLI perform any parsing. they are all a wrapper for getting the information from the CLI command to the cli_bml command.

File: controller/src/beerocks/cli/beerocks_cli_bml.cpp
Comment (adam1985d) [reply controller/src/beerocks/cli/beerocks_cli_bml.cpp]: was changed to bool

File: controller/src/beerocks/cli/beerocks_cli_bml.cpp
Comment (adam1985d) [reply controller/src/beerocks/cli/beerocks_cli_bml.cpp]: ok - replaced the implementation to what you suggested (used the beerocks_cli_socket::client_beacon_11k_req as reference)

File: controller/src/beerocks/cli/beerocks_cli_bml.cpp
Comment (adam1985d) [reply controller/src/beerocks/cli/beerocks_cli_bml.cpp]: wasn't familiar with it - very useful - fixed, thanks

File: controller/src/beerocks/cli/beerocks_cli_bml.cpp
Comment (adam1985d) [reply controller/src/beerocks/cli/beerocks_cli_bml.cpp]: @tomereli already mentioned it - I fixed it and am using a user-defined-sized array declared on the stack.

File: controller/src/beerocks/cli/beerocks_cli_bml.cpp
Comment (adam1985d) [reply controller/src/beerocks/cli/beerocks_cli_bml.cpp]: the code was removed as part of refactoring due to @morantr comment

File: controller/src/beerocks/cli/beerocks_cli_bml.cpp
Comment (adam1985d) [reply controller/src/beerocks/cli/beerocks_cli_bml.cpp]: removed the <calloc> entirely - and will run KW

File: controller/src/beerocks/cli/beerocks_cli_bml.cpp
Comment (mariomaz) [inline controller/src/beerocks/cli/beerocks_cli_bml.cpp]: typo: should be "separated"

File: controller/src/beerocks/cli/beerocks_cli_bml.cpp
Comment (mariomaz) [inline controller/src/beerocks/cli/beerocks_cli_bml.cpp]: typo: "milliseconds"
This typo is repeated in other commands

File: controller/src/beerocks/cli/beerocks_cli_bml.cpp
Comment (mariomaz) [inline controller/src/beerocks/cli/beerocks_cli_bml.cpp]: For the sake of consistency, channel_pool_size and channel_pool should be given in the same order in all commands.

File: controller/src/beerocks/cli/beerocks_cli_bml.cpp
Comment (mariomaz) [reply controller/src/beerocks/cli/beerocks_cli_bml.cpp:1832]: IMO, provided APIs should be always as clear as possible and thus enable should be bool in this case

File: controller/src/beerocks/cli/beerocks_cli_bml.h
Comment (tomereli) [reply controller/src/beerocks/cli/beerocks_cli_bml.h]: I agree with Moran here - we should do the conversion in the BML CLI and get rid of the size of channel pool.

File: controller/src/beerocks/cli/beerocks_cli_bml.h
Comment (tomereli) [reply controller/src/beerocks/cli/beerocks_cli_bml.h]: Agree - the channel_pool should be a vector, you can parse it in the caller.

File: controller/src/beerocks/cli/beerocks_cli_bml.cpp
Comment (mariomaz) [inline controller/src/beerocks/cli/beerocks_cli_bml.cpp]: You should probably be using __func__ instead of a literal string to avoid copy & paste errors. This way maintenance is easier (if method name changes, there is no need to change anything else). 
This is something to be considered in all the methods in this file.

File: controller/src/beerocks/cli/beerocks_cli_bml.cpp
Comment (mariomaz) [inline controller/src/beerocks/cli/beerocks_cli_bml.cpp]: MAC address is not being printed

File: controller/src/beerocks/cli/beerocks_cli_bml.cpp
Comment (mariomaz) [inline controller/src/beerocks/cli/beerocks_cli_bml.cpp]: Method name is being printed together with next literal string

File: controller/src/beerocks/cli/beerocks_cli_bml.cpp
Comment (mariomaz) [inline controller/src/beerocks/cli/beerocks_cli_bml.cpp]: IMO, parentheses in conditionals, even if they are not strictly required, help writing code that is clear and correct

File: controller/src/beerocks/cli/beerocks_cli_bml.cpp
Comment (mariomaz) [inline controller/src/beerocks/cli/beerocks_cli_bml.cpp:1197]: I've been using Yoda conditions for a long time now that these lines make me feel uncomfortable. 
Some static code analyzers would also complain.

**Review by mariomaz — APPROVED** [review]

File: controller/src/beerocks/cli/beerocks_cli_bml.cpp
Comment (adam1985d) [reply controller/src/beerocks/cli/beerocks_cli_bml.cpp]: fixed - thanks

File: controller/src/beerocks/cli/beerocks_cli_bml.cpp
Comment (adam1985d) [reply controller/src/beerocks/cli/beerocks_cli_bml.cpp]: fixed - thanks

File: controller/src/beerocks/cli/beerocks_cli_bml.cpp
Comment (adam1985d) [reply controller/src/beerocks/cli/beerocks_cli_bml.cpp]: I felt the same itch - but this was aligned to the signature of DCS BML APIS.
we decided to remove the channel_pool_size completely (and parse it from the string of channel_pool) - so will be fixed

File: controller/src/beerocks/cli/beerocks_cli_bml.cpp
Comment (adam1985d) [reply controller/src/beerocks/cli/beerocks_cli_bml.cpp]: fixed

File: controller/src/beerocks/cli/beerocks_cli_bml.cpp
Comment (adam1985d) [reply controller/src/beerocks/cli/beerocks_cli_bml.cpp]: fixed

File: controller/src/beerocks/cli/beerocks_cli_bml.cpp
Comment (adam1985d) [reply controller/src/beerocks/cli/beerocks_cli_bml.cpp]: fixed

File: controller/src/beerocks/cli/beerocks_cli_bml.cpp
Comment (adam1985d) [reply controller/src/beerocks/cli/beerocks_cli_bml.cpp]: this is not maintained across the code - but if you find it more readable, I don't mind adding it

fixed.

File: controller/src/beerocks/cli/beerocks_cli_bml.cpp
Comment (adam1985d) [reply controller/src/beerocks/cli/beerocks_cli_bml.cpp:1197]: I am familiar with this approach - but this is not maintained in the prpl.

if @tomereli and @arnout  want to add this requirement to prplmesh guidelines it is OK, but a short search of the following regex. <[a-z,A-Z] == [0-9]> produced 375 results in 93 files

will not be handled in this PR, but we can set a task for this change across the code if decided

File: controller/src/beerocks/cli/beerocks_cli_bml.cpp
Comment (tomereli) [reply controller/src/beerocks/cli/beerocks_cli_bml.cpp]: OK

File: controller/src/beerocks/cli/beerocks_cli_bml.cpp
Comment (tomereli) [reply controller/src/beerocks/cli/beerocks_cli_bml.cpp]: The best would be to change the DCS WLAN APIs as well - is it feasible?

File: controller/src/beerocks/cli/beerocks_cli_bml.cpp
Comment (tomereli) [reply controller/src/beerocks/cli/beerocks_cli_bml.cpp:1197]: Fix this, we will.
image

File: controller/src/beerocks/cli/beerocks_cli_bml.cpp
Comment (mariomaz) [reply controller/src/beerocks/cli/beerocks_cli_bml.cpp]: This was not something to be addressed in this PR but to be discussed if code writing rules are ever revisited.

File: controller/src/beerocks/cli/beerocks_cli_bml.cpp
Comment (adam1985d) [reply controller/src/beerocks/cli/beerocks_cli_bml.cpp]: it might be - you can create a JIRA to check this at a later time - shouldn't block this PR

**Review by tomereli — APPROVED** [review]

File: controller/src/beerocks/cli/beerocks_cli_bml.h
Comment (itayx) [reply controller/src/beerocks/cli/beerocks_cli_bml.h]: Removed `channel_pool_size` throughout

File: controller/src/beerocks/cli/beerocks_cli_bml.h
Comment (itayx) [reply controller/src/beerocks/cli/beerocks_cli_bml.h]: Removed `channel_pool_size` throughout

Comment (adam1985d) [reply controller/src/beerocks/cli/beerocks_cli_bml.h]: fixed - removed the channel pool size from user input params. calculated from channel_pool input.

File: controller/src/beerocks/cli/beerocks_cli_bml.cpp
Comment (adam1985d) [reply controller/src/beerocks/cli/beerocks_cli_bml.cpp]: update - removed the channel_pool_size from user input

**Review by morantr — APPROVED** [review]

File: controller/src/beerocks/cli/beerocks_cli_bml.cpp
Comment (tomereli) [inline controller/src/beerocks/cli/beerocks_cli_bml.cpp:27]: The name of this function implies it creates a string from an array of integers. So it should be 
```
static const std::string string_from_int_array(int arr[], size_t len);
```
Also - please add a doxygen comment explaining this function.

File: controller/src/beerocks/cli/beerocks_cli_bml.cpp
Comment (tomereli) [inline controller/src/beerocks/cli/beerocks_cli_bml.cpp:2044]: I'm think that security mode should be a single value - I'm dazzled why you chose array. Please make it a single value unless there's a very good reason why you need an array. Then we can get rid of this weird `string_from_int_array` function.

File: controller/src/beerocks/cli/beerocks_cli_bml.cpp
Comment (tomereli) [inline controller/src/beerocks/cli/beerocks_cli_bml.cpp:2048]: Same here - there is a single encryption mode set at a given time. I don't understand why it is an array.

File: controller/src/beerocks/cli/beerocks_cli_bml.cpp
Comment (tomereli) [inline controller/src/beerocks/cli/beerocks_cli_bml.cpp:2037]: Instead of this whole big print, why not add `inline std::ostream &operator<<(std::ostream &os, const BML_NEIGHBOR_AP &neigh);` to `struct BML_NEIGHBOR_AP` in `bml_defs.h` then you can replace this with:
```
for (size_t i = 0; i < results_count; i++) {
    std::cout << "result[" << i << "]:" << std::endl << results[i] << std::endl;
}
```

**Review by tomereli — APPROVED** [review]
I added some more comments, I think you should fix them but if this is urgent it can be done as part of #665 , so approving again.

**Comment by tomereli** [comment]
Finally, please rebase and squash before merge.

**Comment by adam1985d** [comment]
fixed all code review comments and passed all tests

**Pull request merged**
